### PR TITLE
P0288R9 move_only_function

### LINKED ISSUE #2241: P0288R9 `move_only_function`
[P0288R9](https://wg21.link/P0288R9) `move_only_function`

Feature-test macro:
`__cpp_lib_move_only_function`

### PR DESCRIPTION
**Pull request by AlexGuteniev**
Resolve #2241

### CODE REVIEW

File: stl/inc/functional
Comment (AdamBucior) [inline stl/inc/functional:34]: Why can't these be inline?

Comment (AlexGuteniev) [reply stl/inc/functional:34]: They will be always called by pointer. So will never actually be inlined.
So I prefer to have them non-inline to reduce the object sizes and linker job.

File: stl/inc/functional
Comment (cpplearner) [inline stl/inc/functional:1237]: `move_only_function::operator()` isn't required to throw `bad_function_call` (and `move_only_function<int() noexcept>::operator()` can't actually throw any exception). Consider just using `nullptr` here.

Comment (AlexGuteniev) [reply stl/inc/functional:1237]: Right, this is _Preconditions_. I would prefer `abort()` then - it will be still easier to debug this way.

File: stl/inc/functional
Comment (MikeGitb) [reply stl/inc/functional:34]: Are  you sure the compiler wouldn't be able to "devirtualize" some of the indirect calls?

File: stl/inc/functional
Comment (StephanTLavavej) [inline stl/inc/functional:1740]: No change requested: `is_pointer_v` isn't specific to function pointers, but object pointers wouldn't be invocable so they can't get here.

Comment (StephanTLavavej) [inline stl/inc/functional:1740]: This doesn't appear to handle N4901 [func.wrap.move.ctor]/8, /8.3: "*Postconditions:* `*this` has no target object if any of the following hold: [...] `remove_cvref_t<F>` is a specialization of the `move_only_function` class template, and `f` has no target object.", which additionally indicates missing test coverage.

This can happen, despite the /5.1 constraint that "`remove_cvref_t<F>` is not the same type as `move_only_function`", when the callable object `F` is a `move_only_function<DifferentSignature>`. For example, constructing `move_only_function<int (int)>` from `move_only_function<short (long long)>`. We have test coverage for `std::function` carefully exercising this case (which you don't have to copy exactly, but can look at for precedent).

File: stl/inc/functional
Comment (StephanTLavavej) [inline stl/inc/functional:1719]: This should be `initializer_list<_Ux>&` according to N4901 [func.wrap.move.ctor]/17.1 "`is_constructible_v<VT, initializer_list<U>&, ArgTypes...>` is `true`, and".

File: tests/std/tests/P0288R9_move_only_function/test.cpp
Comment (StephanTLavavej) [inline tests/std/tests/P0288R9_move_only_function/test.cpp:274]: These lines are repeated. Did you intend to inspect both the source `empty` and the destination `empty_moved`?

File: tests/std/tests/P0288R9_move_only_function/test.cpp
Comment (StephanTLavavej) [inline tests/std/tests/P0288R9_move_only_function/test.cpp:381]: There appears to be no test coverage of construction from `nullptr` (only of construction from null function/member pointers).

Similarly, there's no coverage of assignment from `nullptr`.

Comment (StephanTLavavej) [inline tests/std/tests/P0288R9_move_only_function/test.cpp:381]: There appears to be no test coverage for
```cpp
template<class T, class U, class... Args>
explicit move_only_function(in_place_type_t<T>, initializer_list<U>, Args&&...);
```

Comment (StephanTLavavej) [inline tests/std/tests/P0288R9_move_only_function/test.cpp:381]: There's no test coverage for `result_type`; I recommend verifying that it's properly extracted from all possible function type combinations.

**Review by StephanTLavavej — CHANGES_REQUESTED** [review]

File: stl/inc/functional
Comment (AlexGuteniev) [reply stl/inc/functional:1400]: Probably it was "implicit [[likely]]", though agree that it would be better to swap.

File: stl/inc/functional
Comment (AlexGuteniev) [reply stl/inc/functional:1313]: Probably I should comment this. `nullptr` here is very likely, it means that functor constructor did no throw. I want to avoid dll call of `::operator delete`.

File: stl/inc/functional
Comment (AlexGuteniev) [reply stl/inc/functional:1214]: Probably not by standard, but otherwise the trick works even with other compiler. I'll see if we can avoid unnecessary aliasing

File: stl/inc/functional
Comment (AlexGuteniev) [reply stl/inc/functional:1703]: `_Signature` for `_Fx` is fine, though it has to be `_Signature...` due to another comment, which is slightly more confusing.

File: tests/std/tests/P0288R9_move_only_function/test.cpp
Comment (AlexGuteniev) [reply tests/std/tests/P0288R9_move_only_function/test.cpp:274]: Yes. Now it looks like too paranoid intention to me (check if moved-out empty function is still empty), but I'll follow it.

File: stl/inc/functional
Comment (AlexGuteniev) [reply stl/inc/functional:1247]: I meant "(inner) callable is not present", but "not callable" is also fine

File: stl/inc/functional
Comment (AlexGuteniev) [reply stl/inc/functional:1252]: I didn't bother initially, as it appeared to be tedious, but I'll do it. Probably worth pulling it into the vtable emulation `_Invoke` pointer type either to fully take the advantage of it.

File: stl/inc/functional
Comment (AlexGuteniev) [reply stl/inc/functional:1214]: Strict Aliasing Rule is [[expr.basic.lval]/11](https://eel.is/c++draft/expr#basic.lval-11). It says:

> If a program attempts to access the stored value of an object through a glvalue whose type is not similar to one of the following types the behavior is undefined:
>  * the dynamic type of the object,

_Similar_ is defined in [[conv.qual]/2](https://eel.is/c++draft/expr#conv.qual-2):

> Two types T1 and T2 are similar if they have qualification-decompositions with the same n such that corresponding P components are either the same or one is “array of N” and the other is “array of unknown bound of”, and the types denoted by U are the same.

_Qualification decomposition_ is defined just above in ([conv.qual]/1])[https://eel.is/c++draft/expr#conv.qual-1].

When we apply it, `void*` and `const void*` are _similar types. 

----

Then about _Data. Back to [[expr.basic.lval]/11](https://eel.is/c++draft/expr#basic.lval-11). 11.3 says:
> a char, unsigned char, or std​::​byte type.

----

So nothing here is breaking the rules.
I mean tried to avoid `const void*` and `void*[N]` union, and it turned out to be very inconvenient.
I can try to get rid of `_Data` if you hate it.

File: stl/inc/functional
Comment (AlexGuteniev) [reply stl/inc/functional:1269]: [[expr.basic.lval]/11](https://eel.is/c++draft/expr#basic.lval-11.3), as noted above

File: stl/inc/functional
Comment (AlexGuteniev) [reply stl/inc/functional:1719]: Why? Initializer list is passed by value.

File: stl/inc/functional
Comment (AlexGuteniev) [reply stl/inc/functional:1321]: Commented

File: stl/inc/functional
Comment (StephanTLavavej) [reply stl/inc/functional]: The user's type might care about the difference:

```cpp
#include <initializer_list>
#include <type_traits>
using namespace std;

struct Meow {
    Meow(initializer_list<int>&&, double) = delete;
    Meow(const initializer_list<int>&, double) {}
};

int main() {
    static_assert(!is_constructible_v<Meow, initializer_list<int>, double>);
    static_assert(is_constructible_v<Meow, initializer_list<int>&, double>);
}
```

Comment (AlexGuteniev) [reply stl/inc/functional]: Well, I changed the condition, still not sure why would user's type take `initializer_list` in any other way than by value.

File: stl/inc/functional
Comment (AlexGuteniev) [reply stl/inc/functional:1570]: Ok. But I don't like that in this particular place

File: stl/inc/functional
Comment (AlexGuteniev) [reply stl/inc/functional:1557]: They ask specifically for copying _inv-quals_ [[func.wrap.move.ctor]/1](https://eel.is/c++draft/func.wrap.move#ctor-1),  _inv-quals_ are [[func.wrap.move.general]/2](https://eel.is/c++draft/func.wrap.move#general-2)

I'll change that, but then it would be spelled differently from what we are asked for.

File: stl/inc/functional
Comment (StephanTLavavej) [reply stl/inc/functional]: It's unlikely, but users are permitted to do a lot of wacky things.

File: stl/inc/functional
Comment (StephanTLavavej) [reply stl/inc/functional]: Understood - the Standard itself also generally avoids the `&&`, this is just a rare case where it appears to be implied. I think the difference is acceptable.

File: stl/inc/functional
Comment (StephanTLavavej) [reply stl/inc/functional:1214]: I think this is fine as-is, thanks for explaining!

File: stl/inc/functional
Comment (StephanTLavavej) [reply stl/inc/functional]: Sounds good, thanks!

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
Thanks! I've pushed the following changes (followed by a conflict-free merge with `main`) - please let me know if anything looks wrong.

* Comment spelling/grammar.
* Explicit template args for `_Call` are now unnecessary.
* Comment `#endif // 0`.
* Repair/cleanup generator.
  + **Repair:** Change the `noexcept` machinery from `static constexpr bool _Noexcept` to the latest form.
  + Use `static` for the `constexpr` `inst` array.
  + The raw string literal doesn't contain quotes, so we can simply use `R"()"` instead of `R"--()--"`.
  + De-indent the raw string literal, to match the emitted product code.
  + Also make `ref_str` static.
  + Make `ref_str` `constexpr`. Previously, it was an array of *modifiable* pointers to `const char`.
  + clang-format the generator.
  + Change `noex` to `bool` as it's always treated that way.
  + Also add `const` to all loop variables.
  + Use `const auto` instead of `const char* const` for simplicity.
  + Extract `trait` as it's repeated 3 times. This will also permit line unwrapping.
  + With these changes, the generator's output matches the PR, modulo clang-formatting and a couple of empty lines. (clang-format's behavior can't be easily replicated in the generator, so I concluded it wasn't worth the effort to attempt to match it more closely by avoiding emitting spaces in certain places.)
* Test is already `using namespace std`.
* Switch parameter types to avoid theoretical narrowing.
* Call `test_empty()` in definition order.

**Review by StephanTLavavej — APPROVED** [review]

**Comment by AlexGuteniev** [comment]
Sorry for the mess with generator. I didn't give it that much attention as the product code. (I actually wrote the initial implementation manually, that's why it wasn't present in initial review. But making further changes seemed too tedios to do manually and avoid mistakes)

**Review by barcharcraz — CHANGES_REQUESTED** [review]

File: stl/inc/functional
Comment (AlexGuteniev) [reply stl/inc/functional:1265]: I support this, and I'd like macroizing forwarding via `static_cast`. But I prefer consistent solution in scope of #1208 instead of doing it here once

File: stl/inc/functional
Comment (AlexGuteniev) [reply stl/inc/functional:1368]: Even without RTTI the fake vtable trick was worth doing, since `nullptr` `_Destroy` optimization is helpful.

Comment (AlexGuteniev) [reply stl/inc/functional:1368]: We also do memcpy, which is scary for real vtable

File: stl/inc/functional
Comment (AlexGuteniev) [reply stl/inc/functional:1529]: @barcharcraz: we have expanded version around.
expanded version around:
image

Comment (AlexGuteniev) [reply stl/inc/functional:1529]: Both Go To Definition for `result_type` and interactive debugging into `operator ()` get into just macro.
Well, at least these features are not completely broken by macro, but it is still easier with substituted definition.

File: stl/inc/functional
Comment (StephanTLavavej) [reply stl/inc/functional:1273]: I agree with consistency here (i.e. using `forward` unless and until we decide otherwise). @xiangfan-ms recently informed us that there isn't really a concern for compiler front-end throughput, so debug codegen size is the only remaining concern.

File: stl/inc/functional
Comment (StephanTLavavej) [reply stl/inc/functional]: Although they're tempting, we've generally moved away from macros stamping out runtime code - as @AlexGuteniev noted, they interfere with debugging, among other things. (There are still some occurrences of macros for metaprogramming, notably the horrible architecture-dependent spam for calling conventions, and some of the variant visitor logic - those are less bad because they don't directly emit code that can be stepped into.)

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
I pushed small changes:

* Fixed a comment typo and grammar.
* Added a license banner to the script. (All files need banners unless their format prohibits it.)
* Removed an unnecessary semicolon from the Python. (It was being syntax highlighted red.)

I also verified that the script produces the code in `<functional>` modulo clang-formatting. :heavy_check_mark:

File: stl/inc/functional
Comment (AlexGuteniev) [reply stl/inc/functional]: Excellent catch, thanks!
I did not want to initialize `_Guard_type` with `nullptr`, I instead moved it after `::operator new`. I think this way codegen may be better in case if the functor constructor does not throw.

Comment (StephanTLavavej) [reply stl/inc/functional]: My understanding is that the optimizer is good at eliminating dead stores. Since uninitialized variables are scary, I'd prefer to see a data member initializer to `nullptr`, although because this guard is completely local, I don't think it's absolutely necessary.

File: tests/std/tests/P0288R9_move_only_function/test.cpp
Comment (StephanTLavavej) [inline tests/std/tests/P0288R9_move_only_function/test.cpp:10]: I think we should include Standard `<new>` for `bad_alloc` - unless I'm missing a reason why we need `<vcruntime_new.h>` specifically. (That is, we have a weak preference for writing Standard code in tests - makes things somewhat more maintainable, and might make applying our tests to libc++ easier in the future.)

File: tests/std/tests/P0288R9_move_only_function/test.cpp
Comment (StephanTLavavej) [inline tests/std/tests/P0288R9_move_only_function/test.cpp:506]: We conventionally avoid directly testing integers for being nonzero. (Test code conventions are somewhat weaker than product code, but this one is generally followed, I think).

File: tests/std/tests/P0288R9_move_only_function/test.cpp
Comment (StephanTLavavej) [inline tests/std/tests/P0288R9_move_only_function/test.cpp:516]: `free()` is a guaranteed no-op for null, so we can omit the test here.

**Review by StephanTLavavej — APPROVED** [review]

File: stl/inc/functional
Comment (AlexGuteniev) [reply stl/inc/functional]: Well, I see that on x64 the compiler does not optimize away dead store, but still the effect is small https://godbolt.org/z/qnhehMx9Y

On x86 however where SEH is not free, the cost is easily seen: https://godbolt.org/z/j6EzqdG5c

Comment (AlexGuteniev) [reply stl/inc/functional]: Note that currently the data member is initialized by brace-init. Can add constructor if you wish, can also add initialization of non-member `_Ptr`, this all will not impact the optimization. (I find it hard to explain my compiler intuition, but I checked this statement is correct).

Comment (StephanTLavavej) [reply stl/inc/functional]: Ok, no further change requested. I'll go ahead and resolve this - thanks again @muellerj2, we really appreciate how you caught this bug! :heart_eyes_cat:

File: stl/inc/functional
Comment (barcharcraz) [reply stl/inc/functional]: yeah, OK the generator is fine :D

File: stl/inc/functional
Comment (barcharcraz) [inline stl/inc/functional:1255]: stdcall is not the default calling convention...

**Review by barcharcraz — APPROVED** [review]

File: stl/inc/functional
Comment (StephanTLavavej) [reply stl/inc/functional:1255]: That's correct - it's not the default calling convention, we just need to choose a fixed calling convention that isn't sensitive to `/Gd /Gr /Gz /Gv`.

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

**Comment by StephanTLavavej** [comment]
Thanks for moving the STL's functionality forward to C++23! :joy_cat: :rocket: :tada:

**Comment by MikeGitb** [comment]
Any Idea, when this will be ABI locked? I was considering to play around with the ideas discussed in https://github.com/microsoft/STL/issues/2241 over christmas once there is a merged baseline.

**Comment by AlexGuteniev** [comment]
@MikeGitb , as `C++latest` feature, this is not ABI blocked until C++23 is finalized, and the feature is moved from `C++latest` to `C++23`

**Comment by AlexGuteniev** [comment]
Out of other good news, regarding devirtualization in vtable emulation.
DevCom-1554693 is Fixed in Pending Release; also it did not affect all `/LTGC` cases, only `/LTCG:incremental`.

**Pull request merged**
